rainbow.py

```diff
@@ -263,21 +270,77 @@ def trigger_stream(self):
 
-    def __init__(self, timeout=120, screen=None):
+    def __init__(self, timeout=120, screen=None, joystick=None):
         self.image_width = 320  # Camera image width
         self.image_height = 240  # Camera image height
         self.frame_rate = Fraction(20)  # Camera image capture frame rate
         self.screen = screen
         time.sleep(0.01)
+        self.menu = False
+        self.joystick=joystick
         super(Rainbow, self).__init__(name='Rainbow', timeout=timeout, logger=logger)
+                
+    def setup_controls(self):
+        # colours
+        #why do these need repeating when theyre in menu.py? aren't they global?
+        BLUE = 26, 0, 255
+        SKY = 100, 50, 255
+        CREAM = 254, 255, 250
+        BLACK = 0, 0, 0
+        WHITE = 255, 255, 255
+        control_config = [
+           ("min hue", 5, 90, BLACK, WHITE),
+           ("max hue", 115, 90, BLACK, WHITE),
+           ("min saturation", 5, 165, BLACK, WHITE),
+           ("max saturation", 115, 165, BLACK, WHITE),
+           ("min value", 5, 240, BLACK, WHITE),
+           ("max value", 115, 240, WHITE, WHITE),
+        ]
+        return [
+            self.make_controls(index, *item)
+            for index, item
+            in enumerate(control_config)
+        ]
+   
+
+    def make_controls(self, index, text, xpo, ypo, colour, text_colour):
+        """make a slider control at the specified position"""
+        logger.debug("making button with text '%s' at (%d, %d)", text, xpo, ypo)
+        return dict(
+            index=index,
+            label=text,
+            ctrl = MyScale(label=text, pos=(xpo, ypo), col=colour, min=0, max=255, label_col=text_colour, label_side="top")
+        )
+
+    def joystick_handler(self, button):
+        #if left or right buttons on right side of joystick pressed, treat them like arrow buttons
+        if button['circle']:
+            pygame.event.post(pygame.event.Event(pygame.KEYDOWN,{
+                'mod': 0, 'scancode': 77, 'key': pygame.K_RIGHT, 'unicode': "u'\t'"}))
+        elif button['square']:
+            pygame.event.post(pygame.event.Event(pygame.KEYDOWN,{
+                'mod': 0, 'scancode': 75, 'key': pygame.K_LEFT, 'unicode': "u'\t'"}))
+        elif button['start']:
+            #start button brings up or hides menu
+            self.menu = not self.menu
+            colour = self.processor.colour
+            if not self.menu:
+                #menu closing, store values in file
+                #first, get new values
+                for ctrl in self.controls:
+                    i = ctrl['index']
+                    self.processor.colour_bounds[colour][i % 2][int(i/2)] = ctrl['ctrl'].value
+                data = self.processor.colour_bounds
+                with open('rainbow.json', 'w') as f:
+                    json.dump(data, f)
 
     def run(self):
         # Startup sequence
         logger.info('Setup camera')
+        screen = pygame.display.get_surface()        
```

Comment:
Perhaps a note for future us to investigate why we're needing to get a local `screen`, rather than using the `self.screen` which we set up during `__init__`